core/dbt/events/functions.py

```diff
@@ -129,6 +129,7 @@ def create_text_log_line(e: T_Event, msg_fn: Callable[[T_Event], str]) -> str:
 def create_json_log_line(e: T_Event, msg_fn: Callable[[T_Event], str]) -> str:
     values = e.to_dict(scrub_secrets(msg_fn(e), env_secrets()))
     values['ts'] = e.ts.isoformat()
+    values['data'] = {k: scrub_secrets(str(v), env_secrets()) for (k, v) in e.__dict__.items()}
```

Comment:
This is where my python knowledge is a little weak. I know there are lots of ways to access class attributes and I'm not sure what the tradeoffs of `e.__dict__.items()` is compared to the others. Do you have thoughts on this choice?

Reply:
I should use `vars()` here instead of `__dict__`.  My understanding is there's isn't much of a real difference between the two in terms of what they're doing.  Were you thinking of another way to solve this?

Reply:
not in particular no. I am curious what @iknox-fa thinks though since he might know more of the alternatives.

Reply:
Oh there be dragons there, maybe.  

`vars()` and `_dict_()` both do the same thing, but I'm not sure it's what you want there.  They contain all writeable attributes of an object which means it will be scrubbing some stuff you might not want to scrub (log levels, tags, anything else we hang on an event, etc).

If the scrub method handles that gracefully there's no problems though.

Reply:
`__dict__` only returns the writeable attributes on the subclass.  `inspect.getmembers(e)` shows `__dict__` is just  `{'conn_type': 'postgres', 'conn_name': 'master'}` for the `NewConnection` event.

```@dataclass
class NewConnection(DebugLevel, Cli, File):
    conn_type: str
    conn_name: str

    def message(self) -> str:
        return f'Acquiring new {self.conn_type} connection "{self.conn_name}"'
```

Am I right that since this is just an instance of `NewConnection` that I'm executing on we haven't attached the other writable attributes yet so we're only scrubbing what we care about?

Reply:
> __dict__ only returns the writeable attributes on the subclass

So just to be clear, this isn't generally true, however is here because of the rest of the function as @nathaniel-may points out.

> Am I right that since this is just an instance of NewConnection that I'm executing on we haven't attached the other writable attributes yet so we're only scrubbing what we care about?

Yes, but only by the virtue that we only added fields and all fields are ones we wish to scrub.  Here's an example where it could break stuff.

```
from dataclasses import dataclass, _FIELD_BASE
from pprint import pprint as pp

@dataclass
class NewConnection():
    conn_type: str
    conn_name: str

    def something_unrelated(self):
        self.thing_we_use_somewhere_else = 173

    def message(self) -> str:
        return f'Acquiring new {self.conn_type} connection "{self.conn_name}"'

conn = NewConnection(conn_type="postgres", conn_name="localhost")
conn.something_unrelated()


things_to_scrub = conn.__dict__
pp(things_to_scrub)
# > {'conn_name': 'localhost', 'conn_type': 'postgres','thing_we_use_somewhere_else': 173}
```

So some dev added something to the class and didn't know it was going to be scrubbed and now we try to scrub an int.  No bueno.
Luckily, python gives us some more dunder methods specific to dataclasses we can leverage like so:
```
# continued from above
safer_things_to_scrub = {x:getattr(conn, x) for x,y in conn.__dataclass_fields__.items() if type(y._field_type) == _FIELD_BASE}
pp(safer_things_to_scrub)
# > {'conn_name': 'localhost', 'conn_type': 'postgres'}```

Reply:
So right now we're only scrubbing the message field, not any of the other fields which is going to be a problem once we add these fields. I think we should move scrubbing to the final stringified log line instead. Do you think you could add that to this PR, @emmyoop?

Reply:
@iknox-fa that looks like exactly what we want. Although right now not every class is a dataclass so we might get an `AttributeError` on `x.__dataclass_fields__`. I'm fine making everything a dataclass that might not be currently, as long as it's not abstract because of that mypy bug.

Reply:
@nathaniel-may Oh good point.  You can't use dataclass features without... dataclasses.  This may also run afoul of our 3.6 dataclass library.  I'm def getting into the nitty-gritty of how DCs work and they may not have implemented the full 3.7+ api on those.

Reply:
What about just a simple 
```
if hasattr(e, '__dataclass_fields'):
        values['data2'] = {x:str(getattr(e, x)) for x,y in e.__dataclass_fields__.items() if type(y._field_type) == _FIELD_BASE}
```
We officially deprecated 3.6 (right?) for 1.0.0 so that should be fine.

Reply:
yeah that pattern looks totally reasonable.

Reply:
Yep that works too!